k8s/generate_code.sh

```diff
@@ -1,29 +1,60 @@
 #!/bin/bash
 
-APIS_DIR="github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn"
-VERSION="v1beta1"
-APIS_VERSION_DIR="${APIS_DIR}/${VERSION}"
-OUTPUT_DIR="github.com/longhorn/longhorn-manager/k8s/pkg/client"
-CLIENTSET_DIR="${OUTPUT_DIR}/clientset"
-LISTERS_DIR="${OUTPUT_DIR}/listers"
-INFORMERS_DIR="${OUTPUT_DIR}/informers"
-
-echo Generating deepcopy
-deepcopy-gen --input-dirs ${APIS_VERSION_DIR} \
-	-O zz_generated.deepcopy --bounding-dirs ${APIS_DIR}
-
-echo Generating clientset
-client-gen --clientset-name versioned \
-	--input-base '' --input ${APIS_VERSION_DIR} \
-	--clientset-path ${CLIENTSET_DIR}
-
-echo Generating lister
-lister-gen --input-dirs ${APIS_VERSION_DIR} \
-	--output-package ${LISTERS_DIR}
-
-echo Generating informer
-informer-gen --input-dirs ${APIS_VERSION_DIR} \
-	--versioned-clientset-package "${CLIENTSET_DIR}/versioned" \
-	--listers-package ${LISTERS_DIR} \
-	--output-package ${INFORMERS_DIR}
+set -o errexit
+set -o nounset
+set -o pipefail
 
+LH_MANAGER_DIR="github.com/longhorn/longhorn-manager"
+OUTPUT_DIR="${LH_MANAGER_DIR}/k8s/pkg/client"
+APIS_DIR="${LH_MANAGER_DIR}/k8s/pkg/apis"
+GROUP_VERSION="longhorn:v1beta1"
+CODE_GENERATOR_VERSION="v0.18.0"
+CRDS_DIR="crds"
+CONTROLLER_TOOLS_VERSION="v0.7.0"
+KUSTOMIZE_VERSION="kustomize/v3.10.0"
+
+if [[ -z "${GOPATH}" ]]; then
+  GOPATH=~/go
+fi
+
+if [[ ! -d "${GOPATH}/src/k8s.io/code-generator" ]]; then
+  echo "${GOPATH}/src/k8s.io/code-generator is missing"
+  echo "Prepare to install code-generator"
+	mkdir -p ${GOPATH}/src/k8s.io
+	pushd ${GOPATH}/src/k8s.io
+	git clone -b ${CODE_GENERATOR_VERSION} https://github.com/kubernetes/code-generator 2>/dev/null || true
+	popd
+fi
+
+if ! command -v controller-gen > /dev/null; then
+  echo "controller-gen is missing"
+  echo "Prepare to install controller-gen"
+  go install sigs.k8s.io/controller-tools/cmd/controller-gen@${CONTROLLER_TOOLS_VERSION}
+fi
+
+if ! command -v kustomize > /dev/null; then
+  echo "kustomize is missing"
+  echo "Prepare to install kustomize"
+	mkdir -p ${GOPATH}/src/github.com/kubernetes-sigs
+	pushd ${GOPATH}/src/github.com/kubernetes-sigs
+	git clone -b ${KUSTOMIZE_VERSION} git@github.com:kubernetes-sigs/kustomize.git 2>/dev/null || true
+	cd kustomize/kustomize
+	go install .
+	popd
+fi
+
+${GOPATH}/src/k8s.io/code-generator/generate-groups.sh \
```

Comment:
Avoid unexpected permission error
```suggestion
bash ${GOPATH}/src/k8s.io/code-generator/generate-groups.sh \
```